2-Draft-Accepted/RFCNNNN-Support-Experimental-Features.md

```diff
@@ -0,0 +1,637 @@
+    public ExperimentalFeature[] Dependency { get; }
+}
+```
+
+### Test Experimental Features
+
+The regular daily test pass should run without any experimental feature enabled.
+However, our CI system should be able to run tests with a specific experimental feature turned on,
+so that CI builds can verify PRs for an experimental feature work.
+
+The proposal is to support the tag `[Experiment=<name>]` in our CI system,
+where `<name>` is an experimental feature name.
+When having this tag in the commit message of the last PR commit,
+our CI system will run tests in a PowerShell instance with the experimental feature enabled.
+
+The tests for an experimental feature should be written in one of the following ways:
+
+```powershell
+Describe "Tests for Experimental Feature X" {
+  BeforeAll {
+    $skip = -not $PSEnabledExperimentalFeatures.Contains("X")
+    ...
+  }
+
+  It "..." -Skip:$skip {
+    ...
+  }
+}
+
+try {
+  $skip = -not $PSEnabledExperimentalFeatures.Contains("X")
+  $originalDefaultParameterValues = $PSDefaultParameterValues.Clone()
+  $PSDefaultParameterValues["it:skip"] = $skip
+
+  Describe "Tests for Experimental Feature X" {
+    ...
+  }
+}
+finally {
+  $global:PSDefaultParameterValues = $originalDefaultParameterValues
+}
+```
+
+### Experimental Feature Incompatibility Mitigation
+
+There will be cases where experimental features are incompatible either intentionally or unintentionally.
+
+- An example of an intentional incompatibility occurs when two alternative, but incompatible solutions, are being tested.
+- An example of an unintended incompatibility can occur when two experimental features prove to be incompatible through usage.
+
+A minimal implementation will support defining a list of experimental feature pairs in the `powershell.config.json` file.
+At startup, attempts to use incompatible features would produce warnings and consider both as disabled.
+
+An opportunistic implementation will support an exclusion value in the experimental feature declaration
+for cases where an author is intentionally providing alternative/incompatible implementations.
+
+### Static Code Analysis
+
+Static code analysis support should be provided to address experimental feature authoring.
+Two areas are particularly important:
+
+1. Validating whether an experimental feature is marked correctly.
+   A use case for this occurs when an experimental feature is referenced in a module but not, or incorrectly defined in the manifest.
+   From a script perspective, a `PSScriptAnalyzer` rule would be a reasonable solution.
+   From a compiled code perspective, a reflection-based tool may be needed.
+
+2. Validating whether references to an non-existing experimental feature have been removed.
+   At some point, the experiment is completed and either removed from or incorporated into the code base.
+   At this point, a tool to scan script or compiled code would be needed to ensure any references to the experiment have been removed.
+
```

Comment:
On the static analysis front, I would be in favor of any design choices (like maybe the enum rather than prefixed strings) that would make it easier to run a tool over code. Of course nothing is future proof, but it would be nice to avoid a `"Windows 9"` kind of situation.